### DISCUSSION TITLE
DITA Plugin Development Resources page?

### DISCUSSION THREAD

**Discussion by arkadianriver** [dita-ot/dita-ot#4195] [Q&A]
Hi, all.

I'm an old IBM doc tools person who has found himself looking for work. :roll_eyes: Lately, I've been caught up in the docs-as-code and static-site-generator hype, and have spent less time with DITA. I want to get back into it, however, because it seems to be one of those rare and coveted skills to have these days, similar to "COBOL programming at the time of Y2K." :wink: I say that with love, because I know the value of strongly typed information in some industries, but let's be honest&mdash;XML, XSL, FO, Ant, and Java popularity is waning, despite some powerful CCMS and Authoring tools that are built around them. But, until another technology stack handles typed information as well as these, this is where we're at, right?

Here's my question to the experts (@jelovirt, @infotexture, @robander, ...) and others: The DITA-OT and specification documentation is good at DITA in particular, but **_what's the best way for someone to improve their skills in the technologies that DITA plugin development is built around?_** Do I subscribe to O'Reilly and read the ol' _In a Nutshell_ books on XSL, Ant, Relax NG, etc. from a decade ago, or is there a better way these days&mdash;online courses, certifications, reading each entry in @raducoravu's Oxygen blog ....

I'm very familiar with the DITA spec, have created a few XSL overrides, and have developed an Eclipse user-assistance specialization, but as I peruse the code of some of y'all's excellent DITA plugins on GitHub, I know that I'm rusty and lacking in some areas and could use some guidance. :slightly_smiling_face: Even though this post is self-serving, I do think it might be useful to have a _DITA Development resources_ page in the DITA-OT documentation. Perhaps your answers to this thread could be the start of such a thing? (Or maybe it already exists and I missed it.)

(Sorry for calling each of you out, I meant it mainly as props rather than to demand your attention.. curious about your paths to learning these technologies if you're willing to share, or what you feel might be best-in-class.. the Kernighan & Richies of these technologies.) :sweat_smile:

**Comment by stefan-jung** [comment]
I read the spec, the OT docs, Elliott's book, DITA for Print, and pages 1-3 on Google when searching for "DITA" and "DITA-OT". The relevant information is cluttered and many things are very old. The huge, unattractive technology stack, which you have named, around XML is indeed a problem.

A good starting point for plugin development are the DITA-OT docs. If you are looking for a good starting point for authors, Learning DITA and the Doctales DITA introduction is recommended.

**Comment by raducoravu** [comment]
Indeed when it comes to building plugins the DITA OT docs should have the most recent content:
https://www.dita-ot.org/dev/topics/custom-plugins.html

I also concur that @xephon2 's "Doctales DITA introduction" is worth reading:
https://doctales.github.io/dita-introduction/website/index.html

And if you want to learn by looking at how other people's DITA OT plugins work, we have of open source DITA OT plugins on the Oxygen XML Github account:
https://github.com/search?q=topic%3Adita-ot-plugin+org%3Aoxygenxml&type=Repositories

**Comment by infotexture** [comment]
> it might be useful to have a _DITA Development resources_ page in the DITA-OT documentation

Thanks for this suggestion Gary.

Given our limited resources as a small team of volunteers, it's a bit beyond the scope of the DITA-OT docs to teach all of the skills people need to build their own plug-ins, but I'm open to the idea of adding a topic that points to external resources, like we do in these topics: 
- [Resources for custom PDF plug-ins](https://www.dita-ot.org/dev/topics/pdf-customization-resources.html),
- [DITA-OT Day conference recordings](https://www.dita-ot.org/dev/topics/dita-ot-day-videos.html), and
- [Books](https://www.dita-ot.org/dev/reference/books.html)

Several of the resources mentioned in this discussion are already mentioned in these topics, but it makes sense to collect additional suggestions here.

Though I've not read it myself, I've heard good things about Anderson and Wicentowski’s [XQuery for Humanists](https://www.tamupress.com/book/9781623498290/xquery-for-humanists/), as a book we might recommend.

I have a yellowing copy of DuCharme’s [XSLT Quickly](https://www.manning.com/books/xslt-quickly), but I'm not sure how applicable the 20-year old examples there are to the XSLT code in DITA-OT. The toolkit’s stylesheets have been modernized in certain places over the years to take advantage of performance and usability improvements in XSLT 2 and 3, which were released after the book was published.

I'm not sure if @drmacro’s [specialization tutorials](http://dita4practitioners.github.io/dita-specialization-tutorials/) are up-to-date for recent toolkit versions, but I know people have found them helpful in the past.

I know @xephon2 and others are fond of RELAX NG for specializations, but I'm not sure if there's a concise introduction to the grammar that would be approachable for beginners.

**TL;DR:** Let's keep collecting suggestions here, and distill a shortlist of resources to include in the docs.

**Reply by arkadianriver** [reply]
Thanks so far, everyone! Yes, I started watching the DITA-OT Day recordings a few days ago, very informative! Seems I'll have to plan to enjoy Europe in the Fall!

**Comment by chrispy-snps** [comment]
Some of my favorite XSLT resources:

* _XSLT 2.0 and XPath 2.0 Programmer's Reference_, 4th edition, by Dr. Michael Kay
  * This is my favorite primary reference for learning XSLT and XPath. It strikes a great balance at providing a level of detail approaching the W3C specifications themselves, but written in plainspeak and without the formal presentation.
* [The XSLT mailing list](https://www.mulberrytech.com/xsl/xsl-list/index.html)
  * Even if you don't ask questions, you can learn quite a bit just by following the traffic.
  * The archives are a good source of information too, although it can be tricky to come up with just the right search phrase.
* [XSLT Fiddle](https://xsltfiddle.liberty-development.net/)
  * This is a great site for learning XSLT by interactive experimentation.

I use XSLT 3.0 for everything now; there's no reason not to. I haven't found a solid XSLT 3.0 reference yet; I mostly just search and follow links. I would love to find a book that teaches XSLT/XPath assuming XSLT 3.0 as a baseline, teaching things like maps, arrays, and XPath code blocks as fundamental skills to be applied to stylesheet design.

**Reply by kirkilj** [reply]
There are also a couple XSLT/XPath 3.x online courses at https://www.pluralsight.com/authors/dimitre-novatchev.
